File: bindings/go/download.go
package tiktokdirect

import (
	"context"
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
)

func Download(ctx context.Context, videoURL, kind, output string) (string, error) {
	return New().Download(ctx, videoURL, kind, output)
}

func (e *Extractor) Download(ctx context.Context, videoURL, kind, output string) (string, error) {
	meta, err := e.Extract(ctx, videoURL)
	if err != nil {
		return "", err
	}
	urls := mediaURLs(meta, kind)
	if len(urls) == 0 {
		return "", errors.New("no media URL is available")
	}
	path := outputPath(meta, kind, output)
	for _, mediaURL := range urls {
		if err := e.downloadOne(ctx, mediaURL, path, str(meta["webpage_url"])); err == nil {
			return path, nil
		}
		_ = os.Remove(path)
	}
	return "", errors.New("all media candidates failed")
}

func mediaURLs(meta map[string]any, kind string) []string {
	media := obj(meta, "media")
	music := obj(meta, "music")
	switch kind {
	case "mp4":
		return unique([]string{str(media["play_addr"]), str(media["download_addr"])})
	case "mp3":
		return unique([]string{str(music["play_url"])})
	case "thumbnail":
		return unique([]string{str(meta["thumbnail_url"]), str(media["cover"])})
	default:
		return nil
	}
}

func (e *Extractor) downloadOne(ctx context.Context, url, path, referer string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", e.UserAgent)
	req.Header.Set("Referer", referer)
	resp, err := e.Client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return errors.New(resp.Status)
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()
	_, err = io.Copy(file, resp.Body)
	return err
}

File: bindings/node/src/download.js
import { mkdir, stat, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";

import { unique } from "./values.js";

export async function downloadMedia(extractor, meta, kind, output) {
  const urls = mediaUrls(meta, kind);
  if (!urls.length) throw new Error("no media URL is available");
  const path = await outputPath(meta, kind, output);
  for (const mediaUrl of urls) {
    try {
      const headers = { ...extractor.headers("*/*", meta.webpage_url), range: "bytes=0-" };
      const response = await extractor.fetch(mediaUrl, { headers });
      if (!response.ok) throw new Error(`${response.status} ${response.statusText}`);
      const bytes = Buffer.from(await response.arrayBuffer());
      await mkdir(dirname(path), { recursive: true });
      await writeFile(path, bytes);
      return path;
    } catch {}
  }
  throw new Error("all media candidates failed");
}

function mediaUrls(meta, kind) {
  if (kind === "mp4") {
    return unique([meta.media?.play_addr, meta.media?.download_addr, ...collectNested(meta.raw_item?.video)]);
  }
  if (kind === "mp3") return unique([meta.music?.play_url, ...collectNested(meta.raw_item?.music?.playUrl)]);
  if (kind === "thumbnail") return unique([meta.thumbnail_url, meta.media?.cover]);
  return [];
}

async function outputPath(meta, kind, output = "") {
  if (!output || output.endsWith("/") || output.endsWith("\\")) {
    return join(output || "", defaultFilename(meta, kind));
  }
  try {
    if ((await stat(output)).isDirectory()) return join(output, defaultFilename(meta, kind));
  } catch {}
  return output;
}

function defaultFilename(meta, kind) {
  const ext = kind === "thumbnail" || kind === "avatar" ? "jpg" : kind;
  return `${meta.username || "unknown"}_${meta.video_id || "tiktok"}.${ext}`;
}

function collectNested(value) {
  if (typeof value === "string") return [value];
  if (Array.isArray(value)) return value.flatMap(collectNested);
  if (!value || typeof value !== "object") return [];
  return Object.entries(value).flatMap(([key, item]) => {
    const lower = key.toLowerCase();
    if (key === "bitrateInfo" || lower.endsWith("addr") || lower.endsWith("url") || lower.includes("urllist")) {
      return collectNested(item);
    }
    return [];
  });
}
